src/arch/cortex-m4/ctx_switch.S

```diff
@@ -22,14 +25,51 @@ to_kernel:
   bx lr
 
 .thumb_func
-/* r0 is top of user stack, r1 is heap base */
+/* r0 is top of user stack, r1 Process GOT */
 switch_to_user:
+    push {r4-r11}
+
+    /* Load non-hardware-stacked registers from Process stack */
+    sub r0, #32
+    ldmia r0!, {r4-r11}
     /* Load bottom of stack into Process Stack Pointer */
     msr psp, r0
 
-    push {r4-r11}
+    /* Set PIC base pointer to the Process GOT */
     mov r9, r1
-    svc 0xff
+
+    /* SWITCH */
+    svc 0xff /* It doesn't matter which SVC number we use here */
+
+    /* Push non-harware-stacked registers onto Process stack */
+    /* r0 points to user stack (see to_kernel) */
+    stmdb r0, {r4-r11}
```

Comment:
This is kind of hacky, but no.

Notice the `sub r0, #32` on line 33, right before reading from the stack. We actually store the recent stack as a pointer to where the hardware stacked r0, etc are, rather than where we stored r4-r11. The reason is that the scheduler needs to read and set r0-r3, and this way it can be oblivious to where r7-r11 are stored.

This isn't necessarily the best strategy, but actually stacking these registers in the kernel context is bad anyway since it would be technically possible for a malicious application to use this to write to arbitrary memory. It's easy to see once you realize that actually the process controls the value of r0 in this code (i.e. it's just the process SP).

Instead we should be using a kernel pointer to the process data structure. I'll fix that in a later pull request, since I want to avoid conflicts with other unrelated changes, and then fix it, but the problem is noted on one of the commit messages, and i'll create a tracking issue for it.